shared-module/bitmaptools/__init__.c

```diff
@@ -252,6 +253,123 @@ void common_hal_bitmaptools_fill_region(displayio_bitmap_t *destination,
+
+    mp_obj_t current_point;
+
+
+    size_t tuple_len = 0;
+    mp_obj_t *tuple_items;
+
+    // while there are still points to check
+    while (list_length > 0) {
+        mp_obj_list_get(fill_area, &list_length, &fill_points);
+        current_point = mp_obj_list_pop(fill_area, 0);
+        mp_obj_tuple_get(current_point, &tuple_len, &tuple_items);
+        current_point_color_value = common_hal_displayio_bitmap_get_pixel(
+            destination,
+            mp_obj_get_int(tuple_items[0]),
+            mp_obj_get_int(tuple_items[1]));
+
+        // if the current point is not background color ignore it
+        if (current_point_color_value != replaced_color_value) {
+            mp_obj_list_get(fill_area, &list_length, &fill_points);
+            continue;
+        }
+
+        // fill the current point with fill color
+        displayio_bitmap_write_pixel(
+            destination,
+            mp_obj_get_int(tuple_items[0]),
+            mp_obj_get_int(tuple_items[1]),
+            fill_color_value);
+
+        // add all 4 surrounding points to the list to check
+
+        // ignore points outside of the bitmap
+        if (mp_obj_int_get_checked(tuple_items[1]) - 1 >= 0) {
+            mp_obj_t above_point[] = {
+                tuple_items[0],
+                MP_OBJ_NEW_SMALL_INT(mp_obj_int_get_checked(tuple_items[1]) - 1)
+            };
+            mp_obj_list_append(
+                fill_area,
+                mp_obj_new_tuple(2, above_point));
+        }
+
+        // ignore points outside of the bitmap
+        if (mp_obj_int_get_checked(tuple_items[0]) - 1 >= 0) {
+            mp_obj_t left_point[] = {
+                MP_OBJ_NEW_SMALL_INT(mp_obj_int_get_checked(tuple_items[0]) - 1),
+                tuple_items[1]
+            };
+            mp_obj_list_append(
+                fill_area,
+                mp_obj_new_tuple(2, left_point));
+        }
+
+        // ignore points outside of the bitmap
+        if (mp_obj_int_get_checked(tuple_items[0]) + 1 < destination->width) {
+            mp_obj_t right_point[] = {
+                MP_OBJ_NEW_SMALL_INT(mp_obj_int_get_checked(tuple_items[0]) + 1),
+                tuple_items[1]
+            };
+            mp_obj_list_append(
+                fill_area,
+                mp_obj_new_tuple(2, right_point));
+        }
+
+        // ignore points outside of the bitmap
+        if (mp_obj_int_get_checked(tuple_items[1]) + 1 < destination->height) {
+            mp_obj_t below_point[] = {
+                tuple_items[0],
+                MP_OBJ_NEW_SMALL_INT(mp_obj_int_get_checked(tuple_items[1]) + 1)
+            };
+            mp_obj_list_append(
+                fill_area,
+                mp_obj_new_tuple(2, below_point));
+        }
+
+        mp_obj_list_get(fill_area, &list_length, &fill_points);
+    }
+
+    // set dirty the area so displayio will draw
+    displayio_area_t area = { 0, 0, destination->width, destination->height };
```

Comment:
Were you setting the entire display dirty on purpose? Not saying you need to for this but could track the min/max x and y values to mark only the changed area dirty.

Reply:
Yep I think this is correct it should be efficient to track min/max values as is works and then only make dirty the area that actually did get changed. I'll work on that and then make a new commit once it's complete.

Thanks again for taking a look!